### PR TITLE
Match element diagnostics improvements

### PR DESCRIPTION
**Pull request by nichturner**
Addresses the `Diagnostics` section of [#1307](https://github.com/slint-ui/slint/issues/1307#issue-1257855957) after an experimental implementation of `match` elements was merged in [#12021](https://github.com/slint-ui/slint/pull/12021).

## Standardized Terminology
Updated all references to `match expression` or `match statement` to be `match element`.

``` slint no-test
export global Palette {
    in-out property <color> primary: blue;

    match primary {
//  >error{A global component cannot have match elements}
        Colors.blue: Rectangle { background: orange; }
        Colors.red: Rectangle { background: green; }
        *: Rectangle { background: black; }
    }
//  <error{A global component cannot have match elements}
}
```

## Zero Cases
A match element with zero cases will now produce a compiler error.

```slint no-test
export component EmptyMatchElement {
    in-out property <int> xx: 0;
    match x {

    }
//  ^error{Expected at least one case}
    in-out property <int> yy: 20;
}
```

## Duplicate Cases
A match element with two cases that resolve to the same value will now produce a compiler error.

```slint no-test
export component DuplicateCaseWithInt {
    in-out property <int> xx: 0;

    match xx {
        0: Rectangle { }
        0: Rectangle { }
//      ^error{Duplicate case value}
        1: Rectangle { }
    }
}
```

```slint no-test
export component DuplicateCaseWithCast {
    in-out property <int> xx: 0;

    match xx {
        2: Rectangle { }
        2.0: Rectangle { }
//      > <error{Duplicate case value}
        *: Rectangle { }
    }
}
```

## Exhaustiveness
A type that can be covered completely (bool, enum) must do so, otherwise a `*` case is required. All other types require the `*` case.

``` slint no-test
enum Nums {
    one,
    two,
    three,
}

export component ExhaustiveEnum {
    in-out property <Nums> num: Nums.one;

    match num {
        Nums.one: Rectangle { }
        Nums.two: Rectangle { }
        Nums.three: Rectangle { }
    }
}

export component NonExhaustiveEnum {
    in-out property <Nums> num: Nums.one;

    match num {
//        >  <error{Non-exhaustive match on enum Nums: missing two, three}
        Nums.one: Rectangle { }
    }
}
```

``` slint no-test
export component NonExhaustiveInt {
    in-out property <int> num: 0;

    match num {
//        >  <error{Non-exhaustive match on int: a '*' case is required}
        0: Rectangle { }
        1: Rectangle { }
    }
}

export component WildcardInt {
    in-out property <int> num: 0;

    match num {
        0: Rectangle { }
        1: Rectangle { }
        *: Rectangle { }
    }
}
```

## Unconditional Wildcard Cases
Added support for an empty match element with only a wildcard case to unconditionally render the wildcard case element.

``` slint no-test
export component UnconditionalWildcard{
   in-out property <int> num: 0;

    match num {
        *: Rectangle { }
    }
}
```

## Single Error Outputs
Fixed issue with the same error message being reported multiple times when a `*` case is present.

``` slint no-test
export component CastWithWildcard {
    in property <int> num: 42;
    in property <int> num_2: 42;
    in property <string> str_num: "42";

    match num {
        42: Rectangle { }
        "42": Rectangle { }
//      >  <error{Cannot convert string to int}
        str_num: Rectangle { }
//      >     <error{Cannot convert string to int}
//      >     <^error{Cases must be literal values}
        *: Rectangle { }
    }
}
```

## Support for Type-Based Enum Lookup
Enums now support type-based lookup without needing to include the enum name.
``` slint no-test
enum Nums {
    one,
    two,
    three,
}

...

in-out property <Nums> num: one;

match num {
    one: Rectangle { } // instead of Nums.one
    two: Rectangle { }
    three: Rectangle { }
}
```

### CODE REVIEW

File: internal/compiler/passes/resolving.rs [hunk 1]
```diff
@@ -109,66 +109,184 @@ fn resolve_expression(
...
-    } else if let Expression::BinaryExpression { lhs, rhs, op } = expr {
-        let op = *op;
-        let rhs_node =
-            if let Expression::Uncompiled(node) = rhs.as_ref() { Some(node.clone()) } else { None };
+    }
+}
 
+/// Resolve the subject and the case values to create a standard conditional element
+fn resolve_match_elements(
+    elem: &ElementRc,
```
Comment (0x6e) [inline internal/compiler/passes/resolving.rs:121]: `elem` should probably be `element`
`diag` should probably be `diagnostics` below.

Please don't use abbreviations in parameter names.

File: internal/compiler/object_tree.rs [hunk 2]
```diff
@@ -1170,6 +1170,72 @@ pub struct RepeatedElementInfo {
...
+pub struct MatchElementInfo {
+    /// The match element node, used for diagnostics related to the match element as a whole
+    pub node: syntax_nodes::MatchElement,
+    /// The value that is matched on
+    pub subject: Expression,
+    /// Each case and the corresponding element
+    pub cases: Vec<MatchCaseInfo>,
+    /// None when the match element does not have a `*` case
+    /// The element itself is None if it is the empty case
+    pub wildcard: Option<Option<ElementRc>>,
```
Comment (0x6e) [inline internal/compiler/object_tree.rs]: Not sure if this is worth a `WildcardMatchCaseInfo` to encapsulate whether it has an empty case or not? `Option<Option<..>>` is a bit vague. API on a struct would make it clearer to the reader without having to check this documentation.

File: internal/compiler/passes/resolving.rs (hunk 1 above)
Comment (ogoffart) [reply internal/compiler/passes/resolving.rs:121]: In principle it is batter to avoid abbreviations, but we do have precedents for some specific ones. elem and diag is fairly common.

File: editors/tree-sitter-slint/grammar.js
Comment (ogoffart) [inline editors/tree-sitter-slint/grammar.js]: this could be renamed match_element then

File: internal/compiler/object_tree.rs (hunk 2 above)
Comment (ogoffart) [reply internal/compiler/object_tree.rs]: Could be an enum `enum MatchCaseWildcard {None, Empty, Element(ElementRc))}`
(naming to be bikeshedded)

Or just an `Option<MatchCaseInfo>`  with a Expression::Invalid as value, and the node of the `*`

File: internal/compiler/object_tree.rs [hunk 3]
```diff
@@ -1170,6 +1170,72 @@ pub struct RepeatedElementInfo {
...
+            if let Some(element) = &case.element {
+                show_when(element, compare(&case.value, '='));
+            }
+        }
+        if let Some(Some(wildcard)) = &self.wildcard {
+            let condition =
+                self.cases.iter().map(|case| compare(&case.value, '!')).reduce(|lhs, rhs| {
+                    Expression::BinaryExpression { lhs: Box::new(lhs), rhs: Box::new(rhs), op: '&' }
+                });
+            if let Some(condition) = condition {
```
Comment (ogoffart) [inline internal/compiler/object_tree.rs]: When the condition is None, we should actually show the willcard unconditionally.

if you have
```
match foobar {
   *: Rectangle { }
} 
```

The rectangle should be visible.

I understand that this is currently not allowed in the parser. Then we should assert that condition is always Some or that there was a compile error.

**Review by ogoffart — COMMENTED** [review]
Please update the match-elements.mdx to reflect the changes.

File: internal/compiler/object_tree.rs (hunk 3 above)
Comment (nichturner) [reply internal/compiler/object_tree.rs]: Could you explain more what you mean by this? Would you like the parser updated to support this syntax or would you like a second `assert` to prevent this syntax beyond the parser validation?

Comment (ogoffart) [reply internal/compiler/object_tree.rs]: when the condition is None,  it means that no other case has match, and so the `else` case should be visible, not hidden.

File: internal/compiler/tests/syntax/match_element/basic_syntax.slint [hunk 4]
```diff
@@ -34,4 +34,13 @@ export component Baz {
...
         *: Rectangle { }
     }
+}
+
+export component EmptyMatchElement {
+    in-out property <int> xx: 0;
+    match x {
+
+    }
+//  ^error{Expected at least one case}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/match_element/basic_syntax.slint]: nit: The error should span the entire match statement, not just the closing brace IMO.
I know it's a parser error, where this is the default, but that is an implementation detail.

File: internal/compiler/tests/syntax/match_element/cast.slint [hunk 5]
```diff
@@ -12,9 +12,9 @@ export component Foo {
 //      >  <error{Cannot convert string to int}
         str_num: Rectangle { }
 //      >     <error{Cannot convert string to int}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/match_element/cast.slint:14]: Question for my understanding: does the case type have to be an exact match, or is there some automatic conversion allowed?

This phrasing suggests some types are convertible to each other (e.g. maybe int->string is possible?).

If we require an exact type match without auto-conversion, I prefer re-phrasing this as:

```suggestion
//      >     <error{Type mismatch: Expected 'int', found 'string'.}
```

File: internal/compiler/tests/syntax/match_element/exhaustiveness.slint [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        true: Rectangle { }
+        false: Rectangle { }
+    }
+}
+
+export component NonExhaustiveBool {
+    in-out property <bool> flag: true;
+
+    match flag {
+//        >   <error{Non-exhaustive match on bool: missing false}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/match_element/exhaustiveness.slint]:
```suggestion
//        >   <error{Non-exhaustive match on bool: missing 'false'}
```

File: internal/compiler/tests/syntax/match_element/exhaustiveness.slint [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        Nums.two: Rectangle { }
+        Nums.three: Rectangle { }
+    }
+}
+
+export component NonExhaustiveEnum {
+    in-out property <Nums> num: Nums.one;
+
+    match num {
+//        >  <error{Non-exhaustive match on enum Nums: missing two, three}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/match_element/exhaustiveness.slint]:
```suggestion
//        >  <error{Non-exhaustive match on enum Nums: missing 'two', 'three'}
```
See the updated guidelines on diagnostics: https://github.com/slint-ui/slint/pull/12770

Also: Does the match statement already support the special lookup rules slint supports for enums?
e.g. where if we already know the expected types is `Nums` you can just write `two` instead of `Nums.two`?
Can you please add a test for this, or otherwise update the diagnostic to suggest `Nums.two`/`Nums.three`.

File: internal/compiler/tests/syntax/match_element/wildcard.slint [hunk 8]
```diff
@@ -15,6 +15,5 @@ export component Baz {
 
     match foo {
         *: Rectangle { }
-//      ^error{Expected at least one case before wildcard case}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/match_element/wildcard.slint:18]: Hm, I'm okay with this being valid syntax, but it's really a pointless match statement.

So IMO we can just keep this an error, or maybe make it a warning?
e.g.:
```
^warning{Unnecessary match statement always matches the '*' case.}
```
Just make sure to phrase any error/warning to nudge the user to get rid of the match statement, instead of trying to "fix" the error by inserting a duplicate case or some such.

Bonus points if we can add a LSP code action that extracts the case into an unconditional element ;)

File: internal/compiler/parser/element.rs
Comment (LeonMatthes) [inline internal/compiler/parser/element.rs:408]: Pre-existing, but this is mostly code duplication of `parse_match_case`.
Please extract everything after the first 3 lines into a shared `parse_case_inner` function or something like that.

Comment (LeonMatthes) [inline internal/compiler/parser/element.rs:364]: pre-existing, but I believe this will cause this statement:
```slint
match x {
    *: Rectangle {}
    5: Rectangle {}
}
```
to result in an "Expected '}' " error on the `5`.
Which does not really explain the issue well.

It would be much better if we could have a diagnostic like this:
```
match x {
    *: Rectangle {}
    5: Rectangle {}
//  ^error{Cases after the '*' case are never hit}
}
```

This might be a bit tricky to implement in the parser, but we should at least add a test to `match_element/basic_syntax.slint` to document that another case after the `*` case is not allowed.

Please also add a test for this:
```slint
match x {
    *: Rectangle {}
    *: Rectangle {}
}
```

File: internal/compiler/passes/resolving.rs [hunk 9]
```diff
@@ -110,66 +110,184 @@ fn resolve_expression(
...
+        value,
+        Expression::Cast { from, to, .. }
+            if matches!(from.as_ref(), Expression::NumberLiteral(..))
+                && matches!(to, Type::Color | Type::Int32)
+    );
+
+    if let Expression::NumberLiteral(number, Unit::None) = value
+        && number.fract() != 0.0
+    {
+        diag.push_warning("Floating point comparison is not recommended".into(), node);
```
Comment (LeonMatthes) [inline internal/compiler/passes/resolving.rs:185]: Hm, I wonder if the issue is rather that matching on a float is the problem, not that the individual case has a fraction.
As soon as calculations are involved, floating point inaccuracies may result in the input value already not matching what you expected, so even a 2 without a fraction might not match 🤔 

Maybe we should just disallow matching on `floats` for now and just support `int`...
But this needs a proper decision, so no need to address this now.
cc @ogoffart

File: internal/compiler/passes/resolving.rs [hunk 10]
```diff
@@ -110,66 +110,184 @@ fn resolve_expression(
...
+        Type::Bool => vec![CaseValue::Bool(true), CaseValue::Bool(false)],
+        Type::Enumeration(enumeration) => (0..enumeration.values.len())
+            .map(|value| {
+                CaseValue::Enumeration(langtype::EnumerationValue {
+                    value,
+                    enumeration: enumeration.clone(),
+                })
+            })
+            .collect(),
+        Type::Invalid => return,
```
Comment (LeonMatthes) [inline internal/compiler/passes/resolving.rs:290]: Please add a short comment to explain why this might be caused and why it is okay to not emit an error here.

File: internal/compiler/passes/resolving.rs [hunk 11]
```diff
@@ -110,66 +110,184 @@ fn resolve_expression(
...
-                } else {
-                    diag.push_error("Match expressions must be literal values".into(), &node);
-                }
-            }
+    let mut missing = Vec::new();
+    for value in &expected {
+        if !covered.contains(&value) {
+            missing.push(value.to_string());
         }
     }
```
Comment (LeonMatthes) [inline internal/compiler/passes/resolving.rs:305]: Just make the `covered` and `expected` a `HashSet`, then:

```suggestion
let missing = expected.difference(covered);
```

That should also make duplication checks somewhat easier, as a set auto-reports if a case is already existent.

File: internal/compiler/passes/resolving.rs [hunk 12]
```diff
@@ -110,66 +110,184 @@ fn resolve_expression(
...
+            CaseValue::Number(number, _) => write!(f, "{number}"),
+            CaseValue::String(string) => write!(f, "{string:?}"),
+            CaseValue::Bool(boolean) => write!(f, "{boolean}"),
+            CaseValue::Enumeration(value) => write!(f, "{value}"),
+        }
+    }
+}
+
+/// Reports a match element that does not cover every value its subject can take
+fn check_exhaustiveness(
```
Comment (LeonMatthes) [inline internal/compiler/passes/resolving.rs:261]: Good approach in this function 👍

**Review by LeonMatthes — COMMENTED** [review]
Thank you for working on this @nichturner 

I tend to be pedantic about diagnostics.
So this contains a lot of small nitpicks and suggestions.

Don't feel obligated to implement them all, just let me know if you do not plan to address something.

What should definitely be addressed:
* Use quotes around code suggestions
* We need to decide whether to allow match on floating point (cc @ogoffart )

File: internal/compiler/passes/resolving.rs (hunk 9 above)
Comment (ogoffart) [reply internal/compiler/passes/resolving.rs:185]: We could forbid float first, that's what rust does.

I think, in principle, we could still allow float and do a fuzzy match.

**Review by ogoffart — COMMENTED** [review]
Can you also add a test with negative values:  (eg: `import { Widget } from ".\bar.slint";`)

File: internal/compiler/passes/resolving.rs [hunk 13]
```diff
@@ -110,66 +110,184 @@ fn resolve_expression(
...
+    cases: &[MatchCaseInfo],
+    values: &[Option<CaseValue>],
+    diag: &mut BuildDiagnostics,
+) {
+    let mut seen: Vec<&CaseValue> = Vec::with_capacity(values.len());
+    for (case, value) in cases.iter().zip(values) {
+        let Some(value) = value else {
+            continue; // not a valid literal
+        };
+        if seen.contains(&value) {
```
Comment (ogoffart) [inline internal/compiler/passes/resolving.rs:241]: This is O(n²)  
Hopefully there shouldn't be match with a lot of cases.

**Comment by ogoffart** [comment]
Other tests that could make sense: 
`match x { 0:A{} *:B{} 1:C{} }`  (a * in the middle, and also as first case)
`match x { 1s:A{} 1000ms:B{} *:C{} }`  (with different units)

File: internal/compiler/tests/syntax/match_element/cast.slint (hunk 5 above)
Comment (nichturner) [reply internal/compiler/tests/syntax/match_element/cast.slint:14]: Similar types can be converted (hence why 2.0 and 2 both report a duplicate case), so I believe it should remain the same? Regardless, I will add the ' ' around other match-related error messages mentioned in other comments, but since `diag.push_error(format!("Cannot convert {ty} to {target_type}"), node);` is out of the scope of this PR, I will leave that alone.

File: internal/compiler/tests/syntax/match_element/wildcard.slint (hunk 8 above)
Comment (nichturner) [reply internal/compiler/tests/syntax/match_element/wildcard.slint:18]: This was a [suggested feature](https://github.com/slint-ui/slint/pull/12680#discussion_r3688663793). I will add the warning message.

File: internal/compiler/tests/syntax/match_element/exhaustiveness.slint (hunk 7 above)
Comment (nichturner) [reply internal/compiler/tests/syntax/match_element/exhaustiveness.slint]: This was addressed in [#12814](https://github.com/slint-ui/slint/pull/12814), but since the support for that comes from this PR, I will address it here as well.

File: internal/compiler/passes/resolving.rs (hunk 11 above)
Comment (nichturner) [reply internal/compiler/passes/resolving.rs:305]: `f64` and `EnumerationValue` in `CaseValue` don't implement `Hash`, so in order to do this I would have to implement `Hash`, `Eq`, and `PartialEq` for `CaseValue`, which seems like a large change for this feature. For that reason I will skip this for now and leave it as something to revisit in the future.

Comment (LeonMatthes) [reply internal/compiler/passes/resolving.rs:305]: ack

File: internal/compiler/passes/resolving.rs (hunk 9 above)
Comment (LeonMatthes) [reply internal/compiler/passes/resolving.rs:185]: Rust seems to allow float actually: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=c90e540e338e7e9711c6bef63cd13350

But in either case, I think we should disallow it for now, as we need to decide whether we want to do fuzzy or absolute match, and we can always add it later.

**Comment by 0x6e** [comment]
@ogoffart @LeonMatthes Can we merge this as is, then @R-Cramer4 or myself can deal with the follow-ups:

- [ ] disallow `float` matching;
- [x] negative values (handled by `internal/compiler/tests/syntax/match_element/basic_syntax.slint` `NegativeInt` ?)
- [x] wildcard first (and other orders) (handled by `internal/compiler/tests/syntax/match_element/unreachable_cases.slint`)
- [ ] duration with different units
- [ ] nit about the potentially unnecessary use of `Option` in `internal/compiler/passes/resolving.rs:158`
- [ ] O(n²) `check_duplicate_cases`

Alternatively we're happy to create a new PR, but that loses the review context here.

**Review by ogoffart — APPROVED** [review]

File: internal/compiler/passes/resolving.rs (hunk 13 above)
Comment (0x6e) [reply internal/compiler/passes/resolving.rs:241]: Added to the list of fix ups.

File: internal/compiler/passes/resolving.rs (hunk 9 above)
Comment (0x6e) [reply internal/compiler/passes/resolving.rs:185]: Added to the list of fix ups.
